File: cpp/src/phase2.cpp
#include "simplex/phase2.hpp"
#include <Eigen/LU>
#include <limits>
#include <stdexcept>

SimplexResult
phase2(const LinearProgram& lp, const BasisState& init_basis,
       const PivotRule& rule, const SimplexLogger& logger,
       int maxit, double tol)
{
    const Eigen::MatrixXd& A = lp.A;
    const Eigen::VectorXd& b = lp.b;
    const Eigen::VectorXd& c = lp.c;
    int n = static_cast<int>(A.cols());

    std::vector<int> basic    = init_basis.basic;
    std::vector<int> nonbasic = init_basis.nonbasic;
    int nb = static_cast<int>(nonbasic.size());
    int m  = static_cast<int>(basic.size());

    Eigen::VectorXd x = Eigen::VectorXd::Zero(n);
    {
        Eigen::MatrixXd B(m, m);
        for (int j = 0; j < m; ++j) B.col(j) = A.col(basic[j]);
        Eigen::VectorXd xb_init = B.lu().solve(b);
        for (int j = 0; j < m; ++j) x[basic[j]] = xb_init[j];
    }

    logger.log_phase(2);

    for (int it = 1; it <= maxit; ++it) {
        Eigen::MatrixXd B(m, m), N(m, nb);
        for (int j = 0; j < m;  ++j) B.col(j) = A.col(basic[j]);
        for (int j = 0; j < nb; ++j) N.col(j) = A.col(nonbasic[j]);

        Eigen::PartialPivLU<Eigen::MatrixXd> BF(B);
        Eigen::VectorXd xb    = BF.solve(b);
        Eigen::MatrixXd BinvN = BF.solve(N);

        Eigen::VectorXd cb(m);
        for (int i = 0; i < m; ++i) cb[i] = c[basic[i]];

        Eigen::VectorXd cr(nb);
        for (int j = 0; j < nb; ++j) cr[j] = c[nonbasic[j]];
        cr -= BinvN.transpose() * cb;

        double z = cb.dot(xb);

        logger.log_iteration(it, BasisState{basic, nonbasic}, xb, z);

        auto opt_j = rule.entering_index(cr, tol);
        if (!opt_j.has_value()) {
            x.setZero();
            for (int i = 0; i < m; ++i) x[basic[i]] = xb[i];
            return SimplexResult{ x, z, SolveStatus::OPTIMAL, it };
        }

        int j = *opt_j;
        Eigen::VectorXd d = BinvN.col(j);
        auto opt_i = leaving_index(xb, d, BinvN, tol);

        if (!opt_i.has_value()) {
            Eigen::VectorXd ray = Eigen::VectorXd::Zero(n);
            ray[nonbasic[j]] = 1.0;
            for (int i = 0; i < m; ++i) ray[basic[i]] = -d[i];
            return SimplexResult{ ray, std::numeric_limits<double>::infinity(),
                                  SolveStatus::UNBOUNDED, it };
        }

        int i = *opt_i;
        double theta = xb[i] / d[i];

        x.setZero();
        for (int k = 0; k < m;  ++k) x[basic[k]]    = xb[k] - theta * d[k];
        x[nonbasic[j]] = theta;
        x[basic[i]]    = 0.0;

        std::swap(basic[i], nonbasic[j]);
    }

    throw std::runtime_error("Phase 2 did not converge in " +
                              std::to_string(maxit) +
                              " iterations \xe2\x80\x94 possible cycling");
}

File: rust/src/phase2.rs
use nalgebra::DVector;

use crate::logger::SimplexLogger;
use crate::phase1::col_submatrix;
use crate::pivot::{leaving_index, PivotRule};
use crate::types::{BasisState, LinearProgram, SimplexResult, SolveStatus};

pub fn phase2(
    lp: &LinearProgram,
    init_basis: &BasisState,
    rule: &dyn PivotRule,
    logger: &mut SimplexLogger,
    maxit: usize,
    tol: f64,
) -> SimplexResult {
    let n = lp.ncols();

    let mut basic = init_basis.basic.clone();
    let mut nonbasic = init_basis.nonbasic.clone();

    logger.log_phase(2);

    for it in 1..=maxit {
        let b_mat = col_submatrix(&lp.a, &basic);
        let n_mat = col_submatrix(&lp.a, &nonbasic);

        let lu = b_mat.lu();
        let xb = lu.solve(&lp.b).expect("Phase 2: B is singular");
        let binv_n = lu.solve(&n_mat).expect("Phase 2: B is singular (BinvN)");

        let cb: DVector<f64> = DVector::from_iterator(basic.len(), basic.iter().map(|&i| lp.c[i]));
        let cn: DVector<f64> =
            DVector::from_iterator(nonbasic.len(), nonbasic.iter().map(|&i| lp.c[i]));
        let cr_vec: DVector<f64> = cn - binv_n.transpose() * &cb;
        let z = cb.dot(&xb);

        logger.log_iteration(
            it,
            &BasisState::new(basic.clone(), nonbasic.clone()),
            xb.as_slice(),
            z,
        );

        let j = rule.entering_index(cr_vec.as_slice(), tol);
        if j.is_none() {
            let mut x = DVector::<f64>::zeros(n);
            for (k, &bi) in basic.iter().enumerate() {
                x[bi] = xb[k];
            }
            return SimplexResult { x, z, status: SolveStatus::Optimal, iterations: it };
        }

        let j = j.unwrap();
        let d: Vec<f64> = (0..basic.len()).map(|r| binv_n[(r, j)]).collect();
        let i = leaving_index(xb.as_slice(), &d, &binv_n, tol);

        if i.is_none() {
            let mut ray = DVector::<f64>::zeros(n);
            ray[nonbasic[j]] = 1.0;
            for (k, &bi) in basic.iter().enumerate() {
                ray[bi] = -d[k];
            }
            return SimplexResult { x: ray, z: f64::INFINITY, status: SolveStatus::Unbounded, iterations: it };
        }

        let i = i.unwrap();

        std::mem::swap(&mut basic[i], &mut nonbasic[j]);
    }

    panic!("Phase 2 did not converge in {} iterations — possible cycling", maxit);
}
